org.metaborg.characterclasses/src/main/java/org/metaborg/characterclasses/CharacterClassOptimized.java

```diff
@@ -15,6 +17,8 @@
     private boolean containsEOF; // [256]
 
     private int min, max;
+    // This field is derived from the fields wordX and containsEOF, and is therefore not used in hashCode and equals
+    private boolean empty;
```

Comment:
The `CharacterClassOptimized()` constructor rejects empty character classes, which I think makes sense. We don't need the field and `isEmpty` can just return `false`, right?

Reply:
The `CharacterClassOptimized()` constructor is actually useless, now that I look at it, since it _always_ throws an exception :stuck_out_tongue: I can remove that one.

However, in the constructor `CharacterClassOptimized(long word0, long word1, long word2, long word3, boolean containsEOF, int min, int max)`, no exceptions are thrown if the character class is empty, so in that case the field is still necessary. Or do you think we should actually throw an exception if it is empty?

Reply:
Yes. I think it makes sense to throw an exception in case an empty optimized character is instantiated. At the point of finalizing character classes, parser generation is done and non-empty character classes are useless. I'm not sure if it could happen at all, but actions for empty character classes could be left out the parse table.

Reply:
I've fixed this, and have run the entire build script again to make sure that nothing breaks because of this. I've also added some extra tests.